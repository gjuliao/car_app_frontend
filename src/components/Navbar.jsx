import React, { useCallback, useEffect, useState } from 'react';
import { NavLink, useLocation } from 'react-router-dom';
import { FaFacebookF, FaVimeoV, FaPinterestP } from 'react-icons/fa';
import { AiOutlineTwitter } from 'react-icons/ai';
import { TiSocialGooglePlus } from 'react-icons/ti';
import { useSelector } from 'react-redux';
import getCurrentDate from '../common/utils';
import token from '../redux/Auth/token';
import navbar from '../styles/Navbar.module.css';
import logo from '../assets/images/logo.png';

const handleActive = ({ isActive }) => (isActive
  ? {
    backgroundColor: '#95C010',
    color: 'white',
    textDecoration: 'none',
    transition: '0.5s ease-out',
  }
  : {
    color: 'black',
  });

const Navbar = () => {
  const accessControl = token();
  const [isOpen, setIsOpen] = useState(false);
  const [menuItems, setMenuItems] = useState('');
  const location = useLocation();

  let myUser = localStorage.getItem('user');
  if (myUser) myUser = JSON.parse(myUser);

  const toggleMenu = useCallback(() => {
    setIsOpen(!isOpen);
  }, [isOpen]);

  const handleLogout = () => {
    localStorage.removeItem('token');
    localStorage.removeItem('user');
    window.location.href = '/';
  };
  const user = useSelector((state) => state.session?.data.payload);

  useEffect(() => {
    switch (location.pathname) {
      case '/add-car':
        setMenuItems('');
        break;
      case '/reservation':
        setMenuItems('white-bars');
        break;
      default:
        setMenuItems('');
    }
  }, [location.pathname]);

  return (
    <>
      <div className="sidebar">
        <nav className={navbar.nav}>
          <div role="presentation" className={`${navbar.menu__bars} ${isOpen ? navbar.open : ''}`} onClick={toggleMenu}>
            <div id={`${isOpen ? '' : menuItems}`} className={`${isOpen ? navbar.menu__close : navbar.menu__lines} ${isOpen ? navbar.menubar__firstLine : ''}`} />
            <div id={`${isOpen ? '' : menuItems}`} className={`${isOpen ? navbar.menu__close : navbar.menu__lines} ${isOpen ? navbar.menubar__secondLine : ''}`} />
          </div>
          <div className={`${navbar.menuItems} ${isOpen ? navbar.open : ''}`}>
            <img src={logo} alt="Logo" className={navbar.logo} />
            <NavLink to="/" style={handleActive} className={navbar.link} onClick={toggleMenu}>
              Cars
            </NavLink>
            {
              accessControl ? (
                <>
                  <NavLink to="/reservation" style={handleActive} className={navbar.link} onClick={toggleMenu}>
                    Reservation
                  </NavLink>
                  <NavLink to="/my-reservation" style={handleActive} className={navbar.link} onClick={toggleMenu}>
                    My Reservation
                  </NavLink>
                  {
                  user?.role === 'admin' ? (
                    <>
                      <NavLink to="/add-car" style={handleActive} className={navbar.link} onClick={toggleMenu}>
                        Add Car
                      </NavLink>
                      <NavLink to="/delete-car" style={handleActive} className={navbar.link} onClick={toggleMenu}>
                        Delete Car
                      </NavLink>
                    </>
                  ) : (<></>)

                    }
                  <NavLink to="/" className={navbar.logout} onClick={handleLogout}>
                    Log out
                  </NavLink>
                </>
              ) : (
                <>
                  <NavLink to="/register" className={navbar.signup} onClick={toggleMenu}>
                    Sign Up
                    {' '}
                  </NavLink>
                  <NavLink to="/login" className={navbar.login} onClick={toggleMenu}>
                    Log In
                    {' '}
                  </NavLink>
                </>
              )
            }

            <div className={navbar.nav__footer}>
              <div className={navbar.social__icons}>
                <NavLink to="https://twitter.com/" className={navbar.social_links}>
                  <AiOutlineTwitter />
                </NavLink>
                <NavLink to="https://facebook.com/" className={navbar.social_links}>
                  <FaFacebookF />
                </NavLink>
                <NavLink to="https://www.google.com/" className={navbar.social_links}>
                  {' '}
                  <TiSocialGooglePlus />
                </NavLink>
                <NavLink to="https://www.vimo.me/" className={navbar.social_links}>
                  {' '}
                  <FaVimeoV />
                </NavLink>
                <NavLink to="https://www.pinterest.com/" className={navbar.social_links}>
                  {' '}
                  <FaPinterestP />
                </NavLink>
              </div>
              <div className={navbar.copyright}>
                &copy;
                {getCurrentDate()}
                {' '}
                & Car Reservation
              </div>
            </div>
          </div>
        </nav>
      </div>
    </>
  );
};

export default Navbar;
