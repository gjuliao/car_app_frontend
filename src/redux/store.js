import { configureStore } from '@reduxjs/toolkit';
import reservationReducer from './reservation/reservationSlice';
import sessionReducer from './sessionSlice';
import carListReducer from './carlistSlice';
import carReducer from './carDetail/carDetailSlice';
import carItemsReducer, { fetchCars } from './car/carSlice';
import auth from './Auth/auth';

const store = configureStore({
  reducer: {
    auth,
    reservations: reservationReducer,
    session: sessionReducer,
    cars: carListReducer,
    car: carReducer,
    carlist: carItemsReducer,
  },
});
store.dispatch(fetchCars());
export default store;
