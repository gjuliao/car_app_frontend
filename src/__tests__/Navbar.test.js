import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Navbar from '../components/Navbar';
import store from '../redux/store';

test('Navbar renders without crashing', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Router>
        <Navbar />
      </Router>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
