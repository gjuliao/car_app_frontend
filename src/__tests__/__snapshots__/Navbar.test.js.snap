// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders without crashing 1`] = `
<div
  className="sidebar"
>
  <nav
    className="nav"
  >
    <div
      className="menu__bars "
      onClick={[Function]}
      role="presentation"
    >
      <div
        className="menu__lines "
        id=""
      />
      <div
        className="menu__lines "
        id=""
      />
    </div>
    <div
      className="menuItems "
    >
      <img
        alt="Logo"
        className="logo"
        src=""
      />
      <a
        aria-current="page"
        className="link active"
        href="/"
        onClick={[Function]}
        style={
          {
            "backgroundColor": "#95C010",
            "color": "white",
            "textDecoration": "none",
            "transition": "0.5s ease-out",
          }
        }
      >
        Cars
      </a>
      <a
        className="signup"
        href="/register"
        onClick={[Function]}
      >
        Sign Up
         
      </a>
      <a
        className="login"
        href="/login"
        onClick={[Function]}
      >
        Log In
         
      </a>
      <div
        className="nav__footer"
      >
        <div
          className="social__icons"
        >
          <a
            className="social_links"
            href="https://twitter.com/"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 1024 1024"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M928 254.3c-30.6 13.2-63.9 22.7-98.2 26.4a170.1 170.1 0 0 0 75-94 336.64 336.64 0 0 1-108.2 41.2A170.1 170.1 0 0 0 672 174c-94.5 0-170.5 76.6-170.5 170.6 0 13.2 1.6 26.4 4.2 39.1-141.5-7.4-267.7-75-351.6-178.5a169.32 169.32 0 0 0-23.2 86.1c0 59.2 30.1 111.4 76 142.1a172 172 0 0 1-77.1-21.7v2.1c0 82.9 58.6 151.6 136.7 167.4a180.6 180.6 0 0 1-44.9 5.8c-11.1 0-21.6-1.1-32.2-2.6C211 652 273.9 701.1 348.8 702.7c-58.6 45.9-132 72.9-211.7 72.9-14.3 0-27.5-.5-41.2-2.1C171.5 822 261.2 850 357.8 850 671.4 850 843 590.2 843 364.7c0-7.4 0-14.8-.5-22.2 33.2-24.3 62.3-54.4 85.5-88.2z"
              />
            </svg>
          </a>
          <a
            className="social_links"
            href="https://facebook.com/"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 320 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M279.14 288l14.22-92.66h-88.91v-60.13c0-25.35 12.42-50.06 52.24-50.06h40.42V6.26S260.43 0 225.36 0c-73.22 0-121.08 44.38-121.08 124.72v70.62H22.89V288h81.39v224h100.17V288z"
              />
            </svg>
          </a>
          <a
            className="social_links"
            href="https://www.google.com/"
          >
             
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              version="1"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12.9 13.5l-.7-.5c-.2-.2-.5-.4-.5-.8s.3-.7.6-1c.8-.6 1.7-1.3 1.7-2.8 0-1.5-.9-2.3-1.4-2.7h1.2l1.2-.7h-4.1c-1 0-2.4.2-3.5 1.1-.8.7-1.2 1.7-1.2 2.6 0 1.5 1.2 3.1 3.3 3.1h.6c-.1.2-.2.4-.2.7 0 .6.3 1 .6 1.3-.9.1-2.5.2-3.8.9-1.2.7-1.5 1.7-1.5 2.4 0 1.5 1.4 2.8 4.2 2.8 3.4 0 5.2-1.9 5.2-3.7 0-1.3-.8-1.9-1.7-2.7zm-2.5-2.2c-1.7 0-2.5-2.2-2.5-3.5 0-.5.1-1 .4-1.5.3-.4.9-.7 1.4-.7 1.6 0 2.5 2.2 2.5 3.6 0 .4 0 1-.5 1.4-.3.4-.9.7-1.3.7zm0 7.9c-2.1 0-3.5-1-3.5-2.4s1.3-1.9 1.7-2c.8-.3 1.9-.3 2.1-.3h.5c1.5 1.1 2.1 1.6 2.1 2.6 0 1.2-1 2.1-2.9 2.1zM17 12h-2v-1h2v-1.9l1-.1v2h2v1h-2v2h-1z"
              />
            </svg>
          </a>
          <a
            className="social_links"
            href="https://www.vimo.me/"
          >
             
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M447.8 153.6c-2 43.6-32.4 103.3-91.4 179.1-60.9 79.2-112.4 118.8-154.6 118.8-26.1 0-48.2-24.1-66.3-72.3C100.3 250 85.3 174.3 56.2 174.3c-3.4 0-15.1 7.1-35.2 21.1L0 168.2c51.6-45.3 100.9-95.7 131.8-98.5 34.9-3.4 56.3 20.5 64.4 71.5 28.7 181.5 41.4 208.9 93.6 126.7 18.7-29.6 28.8-52.1 30.2-67.6 4.8-45.9-35.8-42.8-63.3-31 22-72.1 64.1-107.1 126.2-105.1 45.8 1.2 67.5 31.1 64.9 89.4z"
              />
            </svg>
          </a>
          <a
            className="social_links"
            href="https://www.pinterest.com/"
          >
             
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 384 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M204 6.5C101.4 6.5 0 74.9 0 185.6 0 256 39.6 296 63.6 296c9.9 0 15.6-27.6 15.6-35.4 0-9.3-23.7-29.1-23.7-67.8 0-80.4 61.2-137.4 140.4-137.4 68.1 0 118.5 38.7 118.5 109.8 0 53.1-21.3 152.7-90.3 152.7-24.9 0-46.2-18-46.2-43.8 0-37.8 26.4-74.4 26.4-113.4 0-66.2-93.9-54.2-93.9 25.8 0 16.8 2.1 35.4 9.6 50.7-13.8 59.4-42 147.9-42 209.1 0 18.9 2.7 37.5 4.5 56.4 3.4 3.8 1.7 3.4 6.9 1.5 50.4-69 48.6-82.5 71.4-172.8 12.3 23.4 44.1 36 69.3 36 106.2 0 153.9-103.5 153.9-196.8C384 71.3 298.2 6.5 204 6.5z"
              />
            </svg>
          </a>
        </div>
        <div
          className="copyright"
        >
          Â©
          2023
           
          & Car Reservation
        </div>
      </div>
    </div>
  </nav>
</div>
`;
